### PR TITLE
Task-48609: Add the possibility to share a draft with redactors and space managers.

### PR DESCRIPTION
**Pull request by ahmadmido**
Prior to this change, it is not possible to share a draft article with anyone before posting it, the article is private and visible only for the author.
After this change, will ensure that the current author of the draft can select an option (shared or private):
- The share option makes the draft article accessible from news app for all redactors and managers of spaces if the space is redactional, so they can edit the shared draft article.
- The private option makes the draft visible only for the author.
- Only the draft article author is able to change the draft visibility, others can just see the current state.

### CODE REVIEW

File: services/src/main/java/org/exoplatform/news/NewsServiceImpl.java [hunk 1]
```diff
@@ -97,6 +97,10 @@
 
   private final static String      PLATFORM_ADMINISTRATORS_GROUP   = "/platform/administrators";
 
+  public static final String       MIX_NEWS_DRAFT_VISIBILITY              = "mix:draftVisibility";
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/NewsServiceImpl.java]:
```suggestion
  public static final String       NEWS_DRAFT_VISIBILITY_MIXIN_TYPE              = "mix:draftVisibility";
```

File: services/src/main/java/org/exoplatform/news/NewsServiceImpl.java [hunk 2]
```diff
@@ -97,6 +97,10 @@
...
+
+  public static final String       MIX_NEWS_DRAFT_VISIBILITY_PROP         = "exo:draftVisible";
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/NewsServiceImpl.java]:
```suggestion
  public static final String       NEWS_DRAFT_VISIBILE_MIXIN_PROP         = "exo:draftVisible";
```

File: services/src/main/java/org/exoplatform/news/NewsServiceImpl.java [hunk 3]
```diff
@@ -960,6 +973,10 @@ public News createNewsDraft(News news) throws Exception {
     newsDraftNode.setProperty("exo:pinned", false);
     newsDraftNode.setProperty("exo:archived", false);
     newsDraftNode.setProperty("exo:spaceId", news.getSpaceId());
+    if (newsDraftNode.canAddMixin(MIX_NEWS_DRAFT_VISIBILITY)) {
+      newsDraftNode.addMixin(MIX_NEWS_DRAFT_VISIBILITY);
+      newsDraftNode.setProperty(MIX_NEWS_DRAFT_VISIBILITY_PROP, String.valueOf(news.isDraftVisible()));
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/NewsServiceImpl.java]: mixin should be added only once
```suggestion
if (newsDraftNode.canAddMixin(MIX_NEWS_DRAFT_VISIBILITY) && !newsDraftNode.hasProperty(MIX_NEWS_DRAFT_VISIBILITY_PROP)) {
  newsDraftNode.addMixin(MIX_NEWS_DRAFT_VISIBILITY);
}
newsDraftNode.setProperty(MIX_NEWS_DRAFT_VISIBILITY_PROP, String.valueOf(news.isDraftVisible()));
```

File: services/src/main/resources/locale/portlet/news/News_fr.properties [hunk 4]
```diff
@@ -43,6 +43,10 @@ news.composer.modularity.publish=Modalit\u00E9s de diffusion
 news.composer.user.publish=Diffuser l'article \u00E0 tous les utilisateurs
 news.composer.user.publish.info=Cet article sera visible \u00E0 tous les utilisateurs internes m\u00EAme s'ils ne sont pas membres de l'espace
 news.composer.modularity.post=Modalit\u00E9s de publication
+news.composer.shared=Partag\u00E9
+news.composer.private=Priv\u00E9
+news.composer.alert.shared.success=Le brouillon est partag\u00E9 avec les gestionnaires et les r\u00E9dacteurs de l'espace.
```
Comment (azayati) [inline services/src/main/resources/locale/portlet/news/News_fr.properties]:
```suggestion
news.composer.alert.share.draft.success=Le brouillon est partag\u00E9 avec les gestionnaires et les r\u00E9dacteurs de l'espace.
```

File: services/src/main/resources/locale/portlet/news/News_en.properties [hunk 5]
```diff
@@ -43,6 +43,10 @@ news.composer.modularity.publish=Publishing options
 news.composer.user.publish=Publish the article for all users
 news.composer.user.publish.info=This article will be visible for all internal users even if they are not members of the space
 news.composer.modularity.post=Scheduling options
+news.composer.shared= Shared
+news.composer.private=Private
+news.composer.alert.shared.success=The draft is shared with the space's managers and redactors.
```
Comment (azayati) [inline services/src/main/resources/locale/portlet/news/News_en.properties]:
```suggestion
news.composer.alert.share.draft.success=The draft is shared with the space's managers and redactors.
```

File: services/src/main/resources/locale/portlet/news/News_fr.properties [hunk 6]
```diff
@@ -43,6 +43,10 @@ news.composer.modularity.publish=Modalit\u00E9s de diffusion
...
+news.composer.alert.private.success=Ce brouillon n'est plus partag\u00E9 avec les responsables et les r\u00E9dacteurs de l'espace.
```
Comment (azayati) [inline services/src/main/resources/locale/portlet/news/News_fr.properties]:
```suggestion
news.composer.alert.unshare.draft.success=Ce brouillon n'est plus partag\u00E9 avec les responsables et les r\u00E9dacteurs de l'espace.
```

File: services/src/main/resources/locale/portlet/news/News_en.properties [hunk 7]
```diff
@@ -43,6 +43,10 @@ news.composer.modularity.publish=Publishing options
...
+news.composer.alert.private.success=This draft is no longer shared with the space managers and redactors.
```
Comment (azayati) [inline services/src/main/resources/locale/portlet/news/News_en.properties]:
```suggestion
news.composer.alert.unshare.draft.success=This draft is no longer shared with the space managers and redactors.
```

File: webapp/src/main/webapp/components/snackbar/ExoNewsNotificationAlerts.vue [hunk 8]
```diff
@@ -44,6 +44,21 @@ export default {
         });
       }
     });
+    this.$root.$on('update-draft-visibility', status => {
+      if (status) {
+        const message = this.$t('news.composer.alert.shared.success');
```
Comment (azayati) [inline webapp/src/main/webapp/components/snackbar/ExoNewsNotificationAlerts.vue]:
```suggestion
        const message = this.$t('news.composer.alert.share.draft.success');
```

File: webapp/src/main/webapp/components/snackbar/ExoNewsNotificationAlerts.vue [hunk 9]
```diff
@@ -44,6 +44,21 @@ export default {
...
+        this.$root.$emit('news-notification-alert', {
+          message,
+          type: 'info',
+        });
+      } else {
+        const message = this.$t('news.composer.alert.private.success');
```
Comment (azayati) [inline webapp/src/main/webapp/components/snackbar/ExoNewsNotificationAlerts.vue]:
```suggestion
        const message = this.$t('news.composer.alert.unshare.draft.success');
```

File: webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue [hunk 10]
```diff
@@ -36,6 +36,82 @@
           </div>
           <div class="d-flex d-flex flex-grow-0 my-auto">
             <span v-if="!isMobile" class="my-auto me-4 flex-shrink-0">{{ draftSavingStatus }}</span>
+            <div v-if="!editMode && isRedactorSpace">
+              <v-select
+                ref="selectPriority"
```
Comment (azayati) [inline webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue]:
```suggestion
                ref="selectVisibility"
```

File: webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue [hunk 11]
```diff
@@ -291,12 +374,25 @@ export default {
     isMobile() {
       return this.$vuetify.breakpoint.name === 'xs' || this.$vuetify.breakpoint.name === 'sm';
     },
+    isRedactorSpace() {
+      return this.news.id !== '' && this.currentSpace && this.currentSpace.redactorsCount > 0 && this.currentSpace.managersCount > 0;
```
Comment (azayati) [inline webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue]:
```suggestion
      return this.currentSpace && this.currentSpace.redactorsCount > 0;
```

File: webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue [hunk 12]
```diff
@@ -291,12 +374,25 @@ export default {
...
+    },
+    isNewsAuthor() {
+      return  this.news.author === eXo.env.portal.userName;
```
Comment (azayati) [inline webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue]:
```suggestion
      return this.news.author === eXo.env.portal.userName;
```

File: webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue [hunk 13]
```diff
@@ -291,12 +374,25 @@ export default {
...
   },
   watch: {
     'news.title': function() {
       if (this.news.title !== this.originalNews.title) {
         this.autoSave();
       } },
+    'news.draftVisible': function() {
+      if (this.news.draftVisible !== this.originalNews.draftVisible) {
+        this.autoSave();
+      } },
```
Comment (azayati) [inline webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue]:
```suggestion
    'news.draftVisible': function() {
       if (this.news.draftVisible !== this.originalNews.draftVisible) {
         this.autoSave();
       }
     },
```

File: webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue [hunk 14]
```diff
@@ -525,6 +622,8 @@ export default {
             this.news.updater = fetchedNode.updater;
             this.news.draftUpdaterDisplayName = fetchedNode.draftUpdaterDisplayName;
             this.news.draftUpdateDate = fetchedNode.draftUpdateDate;
+            this.news.author = fetchedNode.author;
```
Comment (azayati) [inline webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue:624]: why setting the news author ?

**Review by azayati — CHANGES_REQUESTED** [review]
See review

File: webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue (hunk 11 above)
Comment (ahmadmido) [reply webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue]: this.news.id !== '' it's important to display the selector only if we tab at least text.

File: webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue (hunk 14 above)
Comment (ahmadmido) [reply webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue:624]: We need the information to check isNewsAuthor

File: webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue [hunk 15]
```diff
@@ -375,10 +375,10 @@ export default {
       return this.$vuetify.breakpoint.name === 'xs' || this.$vuetify.breakpoint.name === 'sm';
     },
     isRedactorSpace() {
```
Comment (azayati) [inline webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue]: I suggest to put canUpdateNewVisibilty
return !editMode && this.news.id !== '' && this.currentSpace && this.currentSpace.redactorsCount > 0;

File: webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue [hunk 16]
```diff
@@ -38,7 +38,7 @@
             <span v-if="!isMobile" class="my-auto me-4 flex-shrink-0">{{ draftSavingStatus }}</span>
             <div v-if="!editMode && isRedactorSpace">
```
Comment (azayati) [inline webapp/src/main/webapp/news-activity-composer-app/components/ExoNewsActivityComposer.vue]: div v-if="canUpdateNewVisibilty"

**Review by azayati — CHANGES_REQUESTED** [review]
See review

File: services/src/main/java/org/exoplatform/news/queryBuilder/NewsQueryBuilder.java [hunk 17]
```diff
@@ -33,6 +33,7 @@ public StringBuilder buildQuery(NewsFilter filter) throws Exception {
     org.exoplatform.services.security.Identity currentIdentity = ConversationState.getCurrent().getIdentity();
     boolean isPublisher = currentIdentity.isMemberOf(PLATFORM_WEB_CONTRIBUTORS_GROUP, PUBLISHER_MEMBERSHIP_NAME);
     String username = currentIdentity.getUserId();
+    String spaceId = "exo:spaceId = '";
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/queryBuilder/NewsQueryBuilder.java]: you can define instead exo:spaceId as static final ?

**Review by azayati — CHANGES_REQUESTED** [review]
See review

File: services/src/main/java/org/exoplatform/news/queryBuilder/NewsQueryBuilder.java [hunk 18]
```diff
@@ -19,6 +19,8 @@
 
   private static final String PUBLISHER_MEMBERSHIP_NAME       = "publisher";
 
+  private static final String EXO_SPACE_ID      = "exo:spaceId = '";
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/queryBuilder/NewsQueryBuilder.java]:
```suggestion
  private static final String EXO_SPACE_ID      = "exo:spaceId";
```
you can keep " = '" in the query

**Review by azayati — CHANGES_REQUESTED** [review]
see review

**Review by azayati — APPROVED** [review]
LGTM

**Pull request merged**